lib/srv/server/ssm_install.go

```diff
@@ -84,9 +106,30 @@ func (si *SSMInstaller) Run(ctx context.Context, req SSMRunRequest) error {
 		params[k] = []*string{aws.String(v)}
 	}
 
+	validInstances := ids
+	instancesState, err := si.ssmAgentState(ctx, req, ids)
+	switch {
+	case err != nil:
+		// ssmAgentState uses `ssm:DescribeInstanceInformation` to gather all the instances information.
+		// Previous Docs versions (pre-v16) did not ask for that permission.
+		// If the IAM role does not have access to that action, an Access Denied is returned here.
+		// The process continues but the user is warned that they should add that permission to get better diagnostics.
+		if !trace.IsAccessDenied(err) {
+			return trace.Wrap(err)
+		}
+
+		si.Logger.WarnContext(ctx, "Add ssm:DescribeInstanceInformation action to IAM Role to improve diagnostics of EC2 Teleport installation failures")
+
```

Comment:
Suggestion: Handle AccessDenied directly on the switch? The "double" err check should be negligible.

```suggestion
	case trace.IsAccessDenied(err):
		// ssmAgentState uses `ssm:DescribeInstanceInformation` to gather all the instances information.
		// Previous Docs versions (pre-v16) did not ask for that permission.
		// If the IAM role does not have access to that action, an Access Denied is returned here.
		// The process continues but the user is warned that they should add that permission to get better diagnostics.
		si.Logger.WarnContext(ctx, "Add ssm:DescribeInstanceInformation action to IAM Role to improve diagnostics of EC2 Teleport installation failures")

	case err != nil:
		return trace.Wrap(err)

```